### PR TITLE
[KernelGen][Nvidia] Add embeddingSpMDMGrad operator with Triton kernel

### PR DESCRIPTION
**Pull request by bwbwzzz**
## Summary
Adds a Triton kernel for `embeddingSpMDMGrad`. Triton kernel implementation for embeddingSpMDMGrad.

## Testing
- Parametrized tests over multiple shapes and float16/float32/bfloat16 dtypes (18 cases)
- Validated against PyTorch reference on device side
- Tested on: Nvidia

## Performance
Test command: `pytest benchmark/test_embeddingSpMDMGrad.py --level core` (NVIDIA H20)

### float16
| Size Detail | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|---|---|---|---|
| 2, 4, 8 | 0.009 | 0.009 | 1.067 |
| 2, 4, 8 | 0.009 | 0.010 | 0.930 |
| 4, 8, 32 | 0.009 | 0.009 | 1.021 |
| 4, 8, 32 | 0.009 | 0.010 | 0.983 |
| 1, 3, 64 | 0.009 | 0.009 | 0.935 |
| 1, 3, 64 | 0.010 | 0.009 | 1.112 |

### float32
| Size Detail | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|---|---|---|---|
| 2, 4, 8 | 0.007 | 0.007 | 1.009 |
| 2, 4, 8 | 0.007 | 0.007 | 1.013 |
| 4, 8, 32 | 0.007 | 0.007 | 1.013 |
| 4, 8, 32 | 0.007 | 0.007 | 1.022 |
| 1, 3, 64 | 0.007 | 0.007 | 1.022 |
| 1, 3, 64 | 0.007 | 0.007 | 1.022 |

### bfloat16
| Size Detail | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|---|---|---|---|
| 2, 4, 8 | 0.010 | 0.009 | 1.125 |
| 2, 4, 8 | 0.009 | 0.009 | 1.025 |
| 4, 8, 32 | 0.010 | 0.010 | 1.010 |
| 4, 8, 32 | 0.010 | 0.010 | 0.941 |
| 1, 3, 64 | 0.010 | 0.010 | 1.098 |
| 1, 3, 64 | 0.009 | 0.009 | 1.007 |

**Arithmetic Mean Speedup: 1.020x**
## Multi-backend Testing
| Backend | Accuracy Test | Benchmark | Speedup (mean) | Notes |
|---|---|---|---|---|
| Nvidia (H20) | PASS | PASS (18 cases, --level core) | 1.020 | Primary |
| Tianshu | PASS | PASS (18 cases) | 0.99 | — |
| Muxi | FAIL | FAIL | — | Error: module 'flag_gems' has no attribute 'embeddingSpMDMGrad' |
| Ascend | PASS | — | — | — |
| Hygon | FAIL | PASS (18 cases) | 0.43 | Error: tuple index out of range |

## Files Changed
- `src/flag_gems/ops/embeddingSpMDMGrad.py`: Triton kernel implementation
- `tests/test_embeddingSpMDMGrad.py`: Accuracy test
- `benchmark/test_embeddingSpMDMGrad.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: Math, stage: alpha 5.4)

### CODE REVIEW

File: tests/test_embeddingSpMDMGrad.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    )
+    if padding_idx >= 0 and embeddingsize > 0:
+        mask = torch.rand((Batch, M), device=flag_gems.device) < 0.25
+        indices = torch.where(mask, torch.full_like(indices, padding_idx), indices)
+    num_weights = embeddingsize
+
+    # Reference implementation using embedding_dense_backward
+    ref_grad_output = utils.to_reference(grad_output)
+    ref_indices = utils.to_reference(indices)
+    ref_out = torch.ops.aten.embedding_dense_backward(
```
Comment (w1120029931-bit) [inline tests/test_embeddingSpMDMGrad.py:41]: The benchmark is the embedding_dense_backward operator. Why is it named embeddingSpMDMGrad?

File: tests/test_embeddingSpMDMGrad.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        ref_grad_output,
+        ref_indices,
+        num_weights,
+        padding_idx,
+        False,  # scale_grad_by_freq=False for comparison
+    )
+
+    # GEMS implementation - call directly since embeddingSpMDMGrad is not a standard aten op
+    with flag_gems.use_gems():
+        res_out = flag_gems.embeddingSpMDMGrad(
```
Comment (w1120029931-bit) [inline tests/test_embeddingSpMDMGrad.py:51]: It should be consistent with torch.ops.aten.embedding_dense_backward

File: src/flag_gems/ops/embeddingSpMDMGrad.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        Shape: (num_weights, embed_dim)
+    """
+    logger.debug("GEMS embeddingSpMDMGrad")
+
+    assert indices.dtype in (
+        torch.int32,
+        torch.int64,
+    ), "Indices must be int32 or int64."
+    assert (
+        grad_output.is_cuda and indices.is_cuda and grad_output.device == indices.device
```
Comment (w1120029931-bit) [inline src/flag_gems/ops/embeddingSpMDMGrad.py:77]: **Issue**: CUDA-specific device check found: .is_cuda

**Recommended fix:**
Use flag_gems.device instead of CUDA-specific checks for multi-backend compatibility.

<sub>Rule: `KERNEL_NO_IS_CUDA`</sub>

File: src/flag_gems/ops/embeddingSpMDMGrad.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        indices: Integer indices into the embedding table.
+                 Shape: (N,)
+        num_weights: Total number of embeddings in the table.
+        padding_idx: If specified, the embedding at this index is not updated.
+
+    Returns:
+        Gradient with respect to the embedding weights.
+        Shape: (num_weights, embed_dim)
+    """
+    logger.debug("GEMS embeddingSpMDMGrad")
```
Comment (w1120029931-bit) [inline src/flag_gems/ops/embeddingSpMDMGrad.py:70]: **Suggestion**: Logger message 'GEMS embeddingSpMDMGrad' does not follow 'GEMS <OP_NAME_UPPER>' format

**Recommended fix:**
Use logger.debug("GEMS <OP_NAME_UPPER>") format.

<sub>Rule: `KERNEL_LOGGER_FORMAT`</sub>

File: src/flag_gems/ops/embeddingSpMDMGrad.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    D = grad_output.shape[-1]
+    go = grad_output.contiguous().view(-1, D)  # (N, D)
+    idx = indices.contiguous().view(-1)
+    N = idx.numel()
+
+    assert go.shape[0] == N, "indices number must match grad_output rows."
+    grad_weight_fp32 = torch.zeros((num_weights, D), device=device, dtype=torch.float32)
+
+    BLOCK_D = 128
```
Comment (w1120029931-bit) [inline src/flag_gems/ops/embeddingSpMDMGrad.py:93]: **Suggestion**: Hardcoded hyperparameter without comment: BLOCK_D = 128

**Recommended fix:**
Add a comment explaining why this value was chosen.

<sub>Rule: `HYPERPARAMS`</sub>

**Review by w1120029931-bit — COMMENTED** [review]
## FlagGems Automated Review

**Operator:** `embeddingSpMDMGrad`
**Verdict:** COMMENT

### Summary
- Errors: 1
- Warnings: 2
- Suggestions: 0

File: src/flag_gems/__init__.py [hunk 6]
```diff
@@ -309,6 +309,7 @@ def torch_ge(v):
     ("elu_", elu_),
     ("elu_backward", elu_backward),
     ("embedding", embedding),
+    ("embeddingSpMDMGrad", embeddingSpMDMGrad),
```
Comment (tengqm) [inline src/flag_gems/__init__.py:312]: Is this a torch aten operator?
The error from metax backend gives a negative answer.
